markovflow/models/spatio_temporal_variational.py

```diff
@@ -270,12 +341,256 @@ def posterior(self) -> PosteriorProcess:
+        """
+        Computes the prior distribution on the vector of inducing states
+        """
+        return self._kernel.state_space_model(self._inducing_time)
+
+    def projection_inducing_states_to_observations(self, input_data: tf.Tensor) -> tf.Tensor:
+        """
+        Compute the projection matrix from of the conditional mean of f(x,t) | s(t)
+        :param input_data: Time point and associated spatial dimension to generate observations for,
+         with shape ``batch_shape + [space_dim + 1, num_time_points]``.
+        :return: The projection matrix with shape [num_time_points, obs_dim, num_inducing_time x state_dim ]
+        """
+        inputs, _ = input_data
+        _, inputs_time = inputs[..., :-1], inputs[..., -1]
+        # projection from nearby inducing states to current state (temporal)
+        P, _ = conditional_statistics(inputs_time, self._inducing_time, self._kernel)
+        # projection to state space to observation space (spatial)
+        A = self._kernel.state_to_space_conditional_projection(inputs)
+        return tf.einsum("ncs,nfc->nfs", P, A)
+
+    def update_sites(self, input_data: Tuple[tf.Tensor, tf.Tensor]) -> None:
+        """
+        Perform one joint update of the Gaussian sites
+                𝜽ₘ ← ρ𝜽ₘ + (1-ρ)𝐠ₘ
+
+        Here 𝐠ₘ are the sum of the gradient of the variational expectation for each data point
+        indexed k, projected back to the site vₘ = [uₘ, uₘ₊₁], through the conditional p(fₖ|vₘ)
+        :param input_data: A tuple of time points and observations
+        """
+        inputs, observations = input_data
+        _, inputs_time = inputs[..., :-1], inputs[..., -1]
+
+        fx_mus, fx_covs = self.space_time_predict_f(inputs)
+
+        # get gradient of variational expectations wrt mu, sigma
+        _, grads = self.local_objective_and_gradients(fx_mus, fx_covs, observations)
+
+        P = self.projection_inducing_states_to_observations(input_data)
+        theta_linear, lik_nat2 = back_project_nats(grads[0], grads[1], P)
+
+        # sum sites together
+        indices = tf.searchsorted(self._inducing_time, inputs_time)
+        num_partition = self._inducing_time.shape[0] + 1
+
+        summed_theta_linear = tf.stack(
+            [
+                tf.reduce_sum(l, axis=0)
+                for l in tf.dynamic_partition(theta_linear, indices, num_partitions=num_partition)
+            ]
+        )
+        summed_lik_nat2 = tf.stack(
+            [
+                tf.reduce_sum(l, axis=0)
+                for l in tf.dynamic_partition(lik_nat2, indices, num_partitions=num_partition)
+            ]
+        )
+
+        # update
+        lr = self.learning_rate
+        new_nat1 = (1 - lr) * self.nat1 + lr * summed_theta_linear
+        new_nat2 = (1 - lr) * self.nat2 + lr * summed_lik_nat2
+
+        self.nat2.assign(new_nat2)
+        self.nat1.assign(new_nat1)
+
+    def loss(self, input_data: Tuple[tf.Tensor, tf.Tensor]) -> tf.Tensor:
```

Comment:
Since this method is shared by both subclasses, would it make sense to define this in the base class?